internal/weapons/catalyst/thousandfloatingdreams/thousandfloatingdreams.go

```diff
@@ -0,0 +1,64 @@
+package thousandfloatingdreams
+
+import (
+	"github.com/genshinsim/gcsim/pkg/core"
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/keys"
+	"github.com/genshinsim/gcsim/pkg/core/player/character"
+	"github.com/genshinsim/gcsim/pkg/core/player/weapon"
+	"github.com/genshinsim/gcsim/pkg/modifier"
+)
+
+func init() {
+	core.RegisterWeaponFunc(keys.ThousandFloatingDreams, NewWeapon)
+}
+
+type Weapon struct {
+	Index    int
+	c        *core.Core
+	self     *character.CharWrapper
+	emBonus  float64
+	dmgBonus float64
+	buff     []float64
+}
+
+func (w *Weapon) SetIndex(idx int) { w.Index = idx }
+func (w *Weapon) Init() error {
+	sameCount := 0
+	diffCount := 0
+	for i, char := range w.c.Player.Chars() {
+		if i == w.self.Index {
+			continue
+		}
+		if char.Base.Element == w.self.Base.Element {
+			sameCount++
+		} else {
+			diffCount++
+		}
+	}
+	w.buff[attributes.EM] = w.emBonus * float64(sameCount)
+	w.buff[attributes.EleToDmgP(w.self.Base.Element)] = w.dmgBonus * float64(diffCount)
```

Comment:
Worth nothing that we could derive `diffCount` from `len(w.c.Player.Chars()) - sameCount - 1`. But I am okay leaving it like this since I find it more readable.